### PR TITLE
revamp the data module for better api and working with auto update

### PR DESCRIPTION
**Pull request by tardyp**
In my tests, the midterm version does not work well with updates.

I liked the collection class from the final version, so I include it back, as well as the js query implementation.

The idea is to have the collection listening to its own events and reapply the query when modified.
 some of the details are described here:
https://docs.google.com/document/d/1VRdY2lqyz0UnbD4kllkzXBDIGDDoxvVgmWjn6M3oK4A/edit?ts=563b583f#

* [x] make the base ui work with auto update
* [x] update api doc
* [x] make this work with plugins (console/waterfall)
    * [X] make this work with waterfall
    * [X] make this work with console_view
* [x] make this work with md_base

### CODE REVIEW

**Comment by tardyp** [comment]
@tothandras: intermediate working version. (auto update is not finished though)
If you have any comments on how things are evolving

**Comment by tardyp** [comment]
@rutsky 1b9b86d version has autoupdate working pretty much reliably.
force and rebuild are working, and rebuild does the all the live update until the build finishes.

I want to continue and make a big api change (simplification) before merging it.

dataService.get will not return promise, but directly the collection, as the promise is not appropriate for a constantly updating data

**Comment by tothandras** [comment]
@tardyp You put a lot of work into this! ;) Looks good, although I didn't have time to test it.
> dataService.get will not return promise, but directly the collection, as the promise is not appropriate for a constantly updating data
What about getting related data? In a lot of cases we need to know some resolved data before making the next request.

**Comment by rutsky** [comment]
@tardyp, I checked your recent changes and noticed, that test fails:

```
27 12 2015 00:29:14.827:DEBUG [phantomjs.launcher]: unknown wrapper for Asd

LOG: 'unknown wrapper for', 'Asd'
PhantomJS 1.9.8 (Linux 0.0.0) Data utils service socketPathRE(arg) should return the WebSocket subscribe path of the parameter path FAILED
        Expected '^asd/1/bnm/[^/]+/[^/]+$' to be '^asd\/1\/bnm\/[^/]+\/[^/]+$'.
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44157 <- /source/services/dataUtils/dataUtils.service.spec.coffee:56:12
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44094 <- /source/services/data/data.service.spec.coffee:154:20
            at processQueue (/home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:14792 <- /source/angular.js:14792:0)
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:14808 <- /source/angular.js:14808:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:16052 <- /source/angular.js:16052:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:15870 <- /source/angular.js:15870:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:16160 <- /source/angular.js:16160:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44096 <- /source/services/data/data.service.spec.coffee:149:30
        Expected '^asd/1/[^/]+$' to be '^asd\/1\/[^/]+$'.
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44159 <- /source/services/dataUtils/dataUtils.service.spec.coffee:59:12
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44094 <- /source/services/data/data.service.spec.coffee:154:20
            at processQueue (/home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:14792 <- /source/angular.js:14792:0)
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:14808 <- /source/angular.js:14808:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:16052 <- /source/angular.js:16052:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:15870 <- /source/angular.js:15870:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:16160 <- /source/angular.js:16160:0
            at /home/bob/stuff/buildbot/buildbot/www/data_module/dist/tests.js:44096 <- /source/services/data/data.service.spec.coffee:149:30
```

Other than that issue with test, forcescheduler works for me.

**Comment by rutsky** [comment]
@tardyp, when I open waterfall view (that should contain three completed builds) I get stack overflow error:

```
...
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
equals (vendors.js:10250)
$get.Scope.$digest (vendors.js:24888)
$get.Scope.$apply (vendors.js:25162)
(anonymous function) (d3.service.coffee:10)
fire (vendors.js:3099)
self.fireWith (vendors.js:3211)
done (vendors.js:8264)
(anonymous function) (vendors.js:8605)
```

`equals()` just calls itself recursively.

As I see `equals()` compares `BuildInstance` with other `BuildInstance` (that looks same to me).
The problem is, probably, that `BuildInstance` contains itself in it's field:

```
> o1
BuildInstance {_endpoint: "builds", _accessor: undefined, builderid: 1, buildid: 1, buildrequestid: 1…}
> o1._collection
[BuildInstance, BuildInstance, BuildInstance]
> o1._collection[0] === o1
true
>
```

**Comment by tardyp** [comment]
@tothandras you can see the onNew, onChange callbacks that are made for dealing with loading more data, and doing stuff on the flow.

**Comment by tothandras** [comment]
@tardyp great! :+1:

**Comment by tardyp** [comment]
started the port of md_base.
Having trouble with the build page @shanzi can you help?

**Comment by shanzi** [comment]
Of course. I will have a look at it

**Comment by shanzi** [comment]
@tardyp  It seems that the currently `buildbot/data_module` (whose version is `1.0.10`) is being used instead of `buildbot-data-js` whose version is `1.1.1`, isn't it?

**Comment by tardyp** [comment]
@shanzi, by default it checks out the last published version of data module.

You will need to use the link trick in order to develop with a change data module:
https://github.com/buildbot/buildbot/blob/master/www/data_module/README.md#how-to-test-within-buildbotwwwbase-

File: www/data_module/src/services/data/collection/dataquery.service.coffee [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                        when 'ne' then cmp = v[field] != value
+                        when 'lt' then cmp = v[field] <  value
+                        when 'le' then cmp = v[field] <= value
+                        when 'gt' then cmp = v[field] >  value
+                        when 'ge' then cmp = v[field] >= value
+                        else cmp = v[field] == value or
+                            (angular.isArray(v[field]) and value in v[field]) or
+                            # private fields added by the data service
+                            v["_#{field}"] == value or
+                            (angular.isArray(v["_#{field}"]) and value in v["_#{field}"])
```
Comment (shanzi) [inline www/data_module/src/services/data/collection/dataquery.service.coffee]: @tardyp I found the problem exists here. The `restService` successfully fetched the `builds` of a `builder` but at here they are all be filtered out.

The problems is that, in the query filter, we have a field named `builderid` with its value to be `"2"`(string). But in the array of `BuildInstance`, the `builderid` field is `2`(Integer). The coffee script translate `==` into `===`, so `"2" != 2`. Thus for BuildInstance this function always returns false. This causes the problem that prevent the displaying of `builds`  in `builder` page.

**Comment by shanzi** [comment]
@tardyp I have a trouble setting up `data_modules`'s compiling environment. Gulp says 
```
no such file or directory, scandir '/home/chase/codes/buildbot/www/data_module/node_modules/guanlecoja/lib/types'
```

My node version is `v4.2.4` because `~v4.4.0` not found at NVM. The NPM version is `v2.14.12`

**Comment by rutsky** [comment]
@shanzi I experienced same error, see https://irclogs.jackgrigg.com/irc.freenode.net/buildbot/2015-12-26#i_3767937

**Comment by rutsky** [comment]
Try to install older version of guanlecoja, e.g. `npm install guanlecoja@0.5.0`

File: www/data_module/src/services/data/collection/dataquery.service.coffee (hunk 1 above)
Comment (shanzi) [inline www/data_module/src/services/data/collection/dataquery.service.coffee]: A make a pull request here https://github.com/tardyp/buildbot/pull/7 
Though the list of builds is displaying now, we still have problems displaying a single build. I will look into that later.

**Comment by tardyp** [comment]
Removing do not merge, as I think it is ready for merge.

md_base is "kinda" working, and is in a state were it shows its great potential.
There is still some work to do that I cannot do in a timely manner for this PR.
http://trac.buildbot.net/ticket/3397

I've updated the documentation. A notable change is that I added a dependency on blockdiag, which is a nice sphinx extension to build decent looking diagrams using text only description (kind of like graphviz, but with much better look)

http://interactive.blockdiag.com/?compression=deflate&src=eJzjUgCCpJz85OyUzMR0hWouBTBISSxJ1EtPLXEqzcxJKVbQtVNwzs_JSU0uyczPA_EQ4kkgZcElqQVYVIGFrcFG1nIBAKbsIAc

**Comment by rutsky** [comment]
@tardyp Is waterfall view and console view should work? They don't in my configuration.

Force build and rebuild works good.

**Comment by tardyp** [comment]
Yes, it is supposed to work. what is the error?

Le ven. 1 janv. 2016 à 22:08, Vladimir Rutsky <notifications@github.com> a
écrit :

**Comment by rutsky** [comment]
@tardyp when I open Console view (http://localhost:8080/#/console) I see infinitely spinning loader. I don't see any errors in the Firefox console.

Waterfall view works.

During Force build (and some other operations) I see `ConstraintError` messages in the Firefox console. If it's size constraints error it can be explained by my environment: my laptop has resolution 1366x768 and not all Buildbot UI controls are being fit in my screen.

**Comment by tardyp** [comment]
Console view doesn't work when there is no change source. It only display
bills that are associated to changes. We should display a message when we
detect there are no data to display.

Here, the design make it difficult to distinguish no data compared to data
not yet downloaded. There is a to-do about that


Le sam. 2 janv. 2016 12:51, Vladimir Rutsky <notifications@github.com> a
écrit :

**Comment by tardyp** [comment]
I'd like to merge this by end of day if there are no objection.

**Pull request merged**